crates/ere-risc0/src/prove/bento.rs

```diff
@@ -0,0 +1,230 @@
+    let output = cmd.output()?;
+    if !output.status.success() {
+        return Err(io::Error::other(format!("Failed to run `{cmd:?}`")));
+    }
+    Ok(output)
+}
+
+fn cmd_exec_checked(cmd: &mut Command) -> Result<(), io::Error> {
+    let status = cmd.status()?;
+    if !status.success() {
+        return Err(io::Error::other(format!("Failed to run `{cmd:?}`")));
+    }
+    Ok(())
+}
+
+fn docker_image_exists(image: impl AsRef<OsStr>) -> Result<bool, io::Error> {
+    let output = cmd_output_checked(
+        Command::new("docker")
+            .args(["images", "--quiet"])
+            .arg(image),
+    )?;
+    // If image exists, image id will be printed hence stdout will be non-empty.
+    Ok(!output.stdout.is_empty())
+}
+
+fn docker_image_tag(src: impl AsRef<OsStr>, dst: impl AsRef<OsStr>) -> Result<(), io::Error> {
+    cmd_exec_checked(
+        Command::new("docker")
+            .args(["image", "tag"])
+            .arg(src)
+            .arg(dst),
+    )
+}
+
+pub fn build_bento_images() -> Result<(), io::Error> {
+    let agent_tag = format!("ere-risc0/agent:{SDK_VERSION}");
+    let rest_api_tag = format!("ere-risc0/rest_api:{SDK_VERSION}");
+
+    if docker_image_exists(&agent_tag)? && docker_image_exists(&rest_api_tag)? {
+        return Ok(());
+    }
+
+    let tempdir = tempdir()?;
+
+    cmd_exec_checked(
+        Command::new("git")
+            .args([
+                "clone",
+                "--single-branch",
+                "--branch",
+                &format!("v{SDK_VERSION}"),
+                "https://github.com/risc0/risc0.git",
+            ])
+            .arg(tempdir.path()),
+    )?;
+
+    cmd_exec_checked(
+        Command::new("docker")
+            .arg("compose")
+            .arg("--file")
+            .arg(tempdir.path().join("compose.yml"))
+            .arg("--env-file")
+            .arg(tempdir.path().join("bento/dockerfiles/sample.env"))
+            .arg("build"),
+    )?;
+
+    docker_image_tag("agent", agent_tag)?;
+    docker_image_tag("bento-rest_api", rest_api_tag)?;
+
+    Ok(())
+}
+
+const BENTO_COMPOSE_TEMPLATE: &str = include_str!("../../compose.yml");
+const BENTO_GPU_PROVER_AGENT_TEMPLATE: &str = include_str!("../../gpu_prover_agent.yml");
+
+fn bento_compose_file() -> String {
+    let cuda_visible_devices = env::var("CUDA_VISIBLE_DEVICES").unwrap_or_else(|_| "".to_string());
+    let cuda_visible_devices = cuda_visible_devices
+        .split(",")
+        .flat_map(|device_id| device_id.parse::<usize>().ok())
+        .collect::<Vec<_>>();
+
+    let mut compose: serde_yaml::Mapping = serde_yaml::from_str(BENTO_COMPOSE_TEMPLATE).unwrap();
+    let gpu_prover_agent: serde_yaml::Mapping =
+        serde_yaml::from_str(BENTO_GPU_PROVER_AGENT_TEMPLATE).unwrap();
+
+    if cuda_visible_devices.is_empty() {
+        let mut gpu_prover_agent = gpu_prover_agent.clone();
+        gpu_prover_agent.remove("deploy").unwrap();
```

Comment:
This means this single container prover has access to all GPUs if the `CUDA_VISIBLE_DEVICES` isn't set?

Reply:
In this case if `CUDA_VISIBLE_DEVICES` is not set we assume there is no GPU so we don't give GPU access.

But it seems to make sense to give all GPU by setting `count: all`, since the resource is already assigned to be GPU.